framework/src/main/java/org/tron/core/net/service/effective/ResilienceService.java

```diff
@@ -0,0 +1,170 @@
+      CommonParameter.getInstance().getInactiveThreshold() * 1000L;
+  public static final long blockNotChangeThreshold = 90 * 1000L;
+
+  //when node is isolated, retention percent peers will not be disconnected
+  public static final double retentionPercent = 0.8;
+  private static final int initialDelay = 300;
+  private final String esName = "resilience-service";
+
+  @Autowired
+  private TronNetDelegate tronNetDelegate;
+
+  @Autowired
+  private ChainBaseManager chainBaseManager;
+
+  private ScheduledExecutorService executor;
+
+  public void init() {
+    executor = ExecutorServiceManager.newSingleThreadScheduledExecutor(esName);
+
+    if (Args.getInstance().isOpenFullTcpDisconnect) {
+      executor.scheduleWithFixedDelay(() -> {
+        try {
+          disconnectRandom();
+        } catch (Exception e) {
+          logger.error("DisconnectRandom node failed", e);
+        }
+      }, initialDelay, 60, TimeUnit.SECONDS);
+    } else {
+      logger.info("OpenFullTcpDisconnect is disabled");
+    }
+
+    executor.scheduleWithFixedDelay(() -> {
+      try {
+        disconnectLan();
+      } catch (Exception e) {
+        logger.error("DisconnectLan node failed", e);
+      }
+    }, initialDelay, 10, TimeUnit.SECONDS);
+
+    executor.scheduleWithFixedDelay(() -> {
+      try {
+        disconnectIsolated2();
+      } catch (Exception e) {
+        logger.error("DisconnectIsolated node failed", e);
+      }
+    }, initialDelay, 30, TimeUnit.SECONDS);
+  }
+
+  private void disconnectRandom() {
+    int peerSize = tronNetDelegate.getActivePeer().size();
+    if (peerSize >= CommonParameter.getInstance().getMaxConnections()) {
+      long now = System.currentTimeMillis();
+      List<PeerConnection> peers = tronNetDelegate.getActivePeer().stream()
+          .filter(peer -> !peer.isDisconnect())
+          .filter(peer -> now - peer.getLastActiveTime() >= inactiveThreshold)
+          .filter(peer -> !peer.getChannel().isTrustPeer())
+          .collect(Collectors.toList());
+      if (!peers.isEmpty()) {
+        int index = new Random().nextInt(peers.size());
+        disconnectFromPeer(peers.get(index), ReasonCode.RANDOM_ELIMINATION);
+      }
+    }
+  }
+
+  private void disconnectLan() {
+    if (isLanNode()) {
+      // disconnect from the node that has keep inactive for more than inactiveThreshold
+      // and its lastActiveTime is smallest
+      int peerSize = tronNetDelegate.getActivePeer().size();
+      if (peerSize >= CommonParameter.getInstance().getMinConnections()) {
+        long now = System.currentTimeMillis();
+        Optional<PeerConnection> one = tronNetDelegate.getActivePeer().stream()
+            .filter(peer -> !peer.isDisconnect())
+            .filter(peer -> now - peer.getLastActiveTime() >= inactiveThreshold)
+            .filter(peer -> !peer.getChannel().isTrustPeer())
+            .min(Comparator.comparing(PeerConnection::getLastActiveTime, Long::compareTo));
```

Comment:
This variable `lastActiveTime` has concurrency issues. You can refer to disconnectRandom implementation.

Reply:
Where is disconnectRandom implementation? How to fix it?

Reply:
Add modifier volatile for lastActiveTime to alleviate concurrency issues. Concurrency issues is not very important in this task, it will resume in next cycle.